README
First of all sorry for my clumsy English.

General information:
====================

fprobe-ulog: a NetFlow probe - libipulog-based tool that collect network
traffic data and emit it as NetFlow flows towards the specified
collector.

URL: http://fprobe.sourceforge.net

Compiling and installing:
=========================

Read INSTALL file for basic installation instructions. Below I'll try to
describe advanced compilation options.

--with-membulk=MODE     indexing mode: index8|index16|ptr [default=ptr]
This option concerns to memory management and defines indexing mode and
maximum memory bulk size. I only shall tell, that the `index8' is most
frugal mode, `ptr' - fastest and `index16' somewhere in the middle.

--with-hash=TYPE        hash type: xor8|xor16|crc16 [default=xor16]
fprobe-ulog use hashing to speedup flows cache searching. This option
specifies the hash type. `xor8' is very frugal with memory - it uses
only 1KB (on 32-bit systems) for hash structure while `xor16' and
`crc16' - 256KB. But, on the other hand, bigger hash gives better
performance.
Hash functions xor8 and xor16 faster then crc16, but they are vulnerable
to a DoS attack, as described in "Denial of Service via Algorithmic
Complexity Attacks" by Scott A Crosby and Dan S Wallach:
http://www.cs.rice.edu/~scrosby/hash

--enable-uptime_trick   enable uptime trick [default=yes]
Maybe later...

--enable-icmp_trick     enable icmp trick: yes|cisco|no [default=yes]
If this option set to "yes" fprobe-ulog will store ICMP type and code in
srcport and dstport NetFlow fields.
If this option set to "cisco" only dstport field will used (ICMP type is
the higher eight bits and ICMP code is the lower eight bits). This
storing method used in some Cisco NetFlow implementations.

--enable-debug          enable debugging [default=no]
You may select different events for debugging: (C)apture, (U)npending,
(S)can, (E)mit, (M)emory, (F)ill and (I)nfo. Most interesting (for end
user) from above is Info debugging - you may get general statistic about
captured packets, emitted flows, allocated memory, using kill -s USR1.
Don't forget to run fprobe-ulog with `-v7' otherwise you'll not see
debugging output.
Example:
--enable-debug (enable all events debugging)
--enable-debug=I,C,U,E (enable Info, Capture, Unpending and Emit debugging)

Brief explanation of Info debug messages:
I: received:[total packets]/[fragmented] ([total size])
   pending:[now in queue]/[maximum]
I: ignored:[non-IP] lost:[pending queue full]+[no memory for caching]
I: cache:[flows]/[fragmented] emit:[sent datagrams]/[sent flows]/[flows
   in emit queue]
I: memory:[allocated flows]/[free] ([allocated memory in bytes])

--enable-messages       enable runtime messages [default=no]
This option enables non-fatal runtime errors reporting. Be carefull - it
may flood your syslog.

--with-piddir=DIR       pidfiles location [default=/var/run]
Directory to store pidfiles.

Useful links:
=============

nProbe - NetFlow probe by Luca Deri:
http://www.ntop.org/nProbe.html

fprobe (namesake of mine project) - NetFlow probe by Bogdan Surdu:
http://psi.home.ro/flow

Softflowd - traffic analyzer capable of Cisco NetFlow data export:
http://www.mindrot.org/softflowd.html

Cisco's NetFlow links:
http://www.cisco.com/go/netflow

Excellent links collection about Network Monitoring and Analysis:
http://www.switch.ch/tf-tant/floma

Contacts:
=========

Feel free to send any questions, comments, bug reports etc.
Contributions are welcome, including cosmetic fixes and pointing out
usability problems.

Sincerely yours,
Slava Astashonok <sla@0n.ru>


ChangeLog
2014-12-23  Orion Poplawski <orion@cora.nwra.com>

	* Release 1.2: Convert to libnetfilter_log_libipulog

2005-01-30  Slava Astashonok <sla@0n.ru>

	* Release 1.1: bugfixes, features enhancement.

	* Changes:
	- was fixed vulnerability to a DoS attack, as described in "Denial
          of Service via Algorithmic Complexity Attacks" by Scott A
          Crosby and Dan S Wallach: http://www.cs.rice.edu/~scrosby/hash,
	  by strengthening the crc16 hash function implementation;
	  note that xor8 and xor16 hashes remain vulnerable
	- default hash was changed to crc16
	- -B and -M options were added
	- if_nametoindex() used as default action to find SNMP-index

2004-09-21  Slava Astashonok <sla@0n.ru>

	* Release 1.0.6: minor bugfixes.

	* Changes:
	- security fix for "change user" feature
	- typo in CAPTURE_SIZE definition was fixed
	- debian/* files were removed

2004-08-14  Slava Astashonok <sla@0n.ru>

	* Release 1.0.5: initial ULOG fork


src/netflow.h
/*
	Copyright (C) Slava Astashonok <sla@0n.ru>

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License.

	$Id: netflow.h,v 1.3.2.5.2.1 2004/08/13 21:03:10 sla Exp $
*/

#ifndef _NETFLOW_H_
#define _NETFLOW_H_

#include <my_inttypes.h>

#define NETFLOW_BYTES_32 1
#define NETFLOW_BYTES_32_SIZE 4

#define NETFLOW_PKTS_32 2
#define NETFLOW_PKTS_32_SIZE 4

#define NETFLOW_FLOWS 3
#define NETFLOW_FLOWS_SIZE 4

#define NETFLOW_PROT 4
#define NETFLOW_PROT_SIZE 1

#define NETFLOW_SRC_TOS 5
#define NETFLOW_SRC_TOS_SIZE 1

#define NETFLOW_TCP_FLAGS 6
#define NETFLOW_TCP_FLAGS_SIZE 1

#define NETFLOW_L4_SRC_PORT 7
#define NETFLOW_L4_SRC_PORT_SIZE 2

#define NETFLOW_IPV4_SRC_ADDR 8
#define NETFLOW_IPV4_SRC_ADDR_SIZE 4

#define NETFLOW_SRC_MASK 9
#define NETFLOW_SRC_MASK_SIZE 1

#define NETFLOW_INPUT_SNMP 10
#define NETFLOW_INPUT_SNMP_SIZE 2

#define NETFLOW_L4_DST_PORT 11
#define NETFLOW_L4_DST_PORT_SIZE 2

#define NETFLOW_IPV4_DST_ADDR 12
#define NETFLOW_IPV4_DST_ADDR_SIZE 4

#define NETFLOW_DST_MASK 13
#define NETFLOW_DST_MASK_SIZE 1

#define NETFLOW_OUTPUT_SNMP 14
#define NETFLOW_OUTPUT_SNMP_SIZE 2

#define NETFLOW_IPV4_NEXT_HOP 15
#define NETFLOW_IPV4_NEXT_HOP_SIZE 4

#define NETFLOW_SRC_AS 16
#define NETFLOW_SRC_AS_SIZE 2

#define NETFLOW_DST_AS 17
#define NETFLOW_DST_AS_SIZE 2

#define NETFLOW_BGP_NEXT_HOP 18
#define NETFLOW_BGP_NEXT_HOP_SIZE 4

#define NETFLOW_IPM_DPKTS 19
#define NETFLOW_IPM_DPKTS_SIZE 4

#define NETFLOW_IPM_DOCTETS 20
#define NETFLOW_IPM_DOCTETS_SIZE 4

#define NETFLOW_LAST_SWITCHED 21
#define NETFLOW_LAST_SWITCHED_SIZE 4

#define NETFLOW_FIRST_SWITCHED 22
#define NETFLOW_FIRST_SWITCHED_SIZE 4

#define NETFLOW_BYTES_64 23
#define NETFLOW_BYTES_64_SIZE 8

#define NETFLOW_PKTS_64 24
#define NETFLOW_PKTS_64_SIZE 8

#define NETFLOW_MAC_ADDR 25
#define NETFLOW_MAC_ADDR_SIZE 6

#define NETFLOW_VLAN_ID 26
#define NETFLOW_VLAN_ID_SIZE 2

#define NETFLOW_IPV6_SRC_ADDR 27
#define NETFLOW_IPV6_SRC_ADDR_SIZE 16

#define NETFLOW_IPV6_DST_ADDR 28
#define NETFLOW_IPV6_DST_ADDR_SIZE 16

#define NETFLOW_IPV6_SRC_MASK 29
#define NETFLOW_IPV6_SRC_MASK_SIZE 1

#define NETFLOW_IPV6_DST_MASK 30
#define NETFLOW_IPV6_DST_MASK_SIZE 1

#define NETFLOW_FLOW_LABEL 31
#define NETFLOW_FLOW_LABEL_SZIE 3

#define NETFLOW_ICMP_TYPE 32
#define NETFLOW_ICMP_TYPE_SIZE 1

#define NETFLOW_MUL_IGMP_TYPE 33
#define NETFLOW_MUL_IGMP_TYPE_SIZE 1

#define NETFLOW_SAMPLING_INTERVAL 34
#define NETFLOW_SAMPLING_INTERVAL_SIZE 4

#define NETFLOW_SAMPLING_ALGORITHM 35
#define NETFLOW_SAMPLING_ALGORITHM_SIZE 1

#define NETFLOW_FLOW_ACTIVE_TIMEOUT 36
#define NETFLOW_FLOW_ACTIVE_TIMEOUT_SIZE 2

#define NETFLOW_FLOW_INACTIVE_TIMEOUT 37
#define NETFLOW_FLOW_INACTIVE_TIMEOUT_SIZE 2

#define NETFLOW_ENGINE_TYPE 38
#define NETFLOW_ENGINE_TYPE_SIZE 1

#define NETFLOW_ENGINE_ID 39
#define NETFLOW_ENGINE_ID_SIZE 1

#define NETFLOW_TOTAL_BYTES_EXPORTED 40
#define NETFLOW_TOTAL_BYTES_EXPORTED_SIZE 4

#define NETFLOW_TOTAL_EXPORT_PKTS_SENT 41
#define NETFLOW_TOTAL_EXPORT_PKTS_SENT_SIZE 4

#define NETFLOW_TOTAL_FLOWS_EXPORTED 42
#define NETFLOW_TOTAL_FLOWS_EXPORTED_SIZE 4

#define NETFLOW_FLOW_SAMPLER_ID 48
#define NETFLOW_FLOW_SAMPLER_ID_SIZE 1

#define NETFLOW_FLOW_SAMPLER_MODE 49
#define NETFLOW_FLOW_SAMPLER_MODE_SIZE 1

#define NETFLOW_FLOW_SAMPLER_RANDOM_INTERVAL 50
#define NETFLOW_FLOW_SAMPLER_RANDOM_INTERVAL_SIZE 4

#define NETFLOW_IP_PROTOCOL_VERSION 60
#define NETFLOW_IP_PROTOCOL_VERSION_SIZE 1

#define NETFLOW_DIRECTION 61
#define NETFLOW_DIRECTION_SIZE 1

#define NETFLOW_IPV6_NEXT_HOP 62
#define NETFLOW_IPV6_NEXT_HOP_SIZE 16

#define NETFLOW_BGP_IPV6_NEXT_HOP 63
#define NETFLOW_BGP_IPV6_NEXT_HOP_SIZE 16

#define NETFLOW_IPV6_OPTION_HEADERS 64
#define NETFLOW_IPV6_OPTION_HEADERS_SIZE 4


#define NETFLOW_VERSION 1001
#define NETFLOW_VERSION_SIZE 2

#define NETFLOW_COUNT 1002
#define NETFLOW_COUNT_SIZE 2

#define NETFLOW_UPTIME 1003
#define NETFLOW_UPTIME_SIZE 4

#define NETFLOW_UNIX_SECS 1004
#define NETFLOW_UNIX_SECS_SIZE 4

#define NETFLOW_UNIX_NSECS 1005
#define NETFLOW_UNIX_NSECS_SIZE 4

#define NETFLOW_FLOW_SEQUENCE 1006
#define NETFLOW_FLOW_SEQUENCE_SIZE 4

#define NETFLOW_HEADER_SAMPLING_INTERVAL 1009 /* ??? */
#define NETFLOW_HEADER_SAMPLING_INTERVAL_SIZE 2

#define NETFLOW_FLAGS7_1 1010
#define NETFLOW_FLAGS7_1_SIZE 1

#define NETFLOW_FLAGS7_2 1011
#define NETFLOW_FLAGS7_2_SIZE 2

#define NETFLOW_ROUTER_SC 1012
#define NETFLOW_ROUTER_SC_SIZE 4


#define NETFLOW_PAD8 9001
#define NETFLOW_PAD8_SIZE 1

#define NETFLOW_PAD16 9002
#define NETFLOW_PAD16_SIZE 2

#define NETFLOW_PAD32 9003
#define NETFLOW_PAD32_SIZE 4

#define NETFLOW_PAD64 9004
#define NETFLOW_PAD64_SIZE 8


#define NETFLOW_MAX_PACKET 1500

#define NETFLOW1_VERSION 1
#define NETFLOW1_HEADER_SIZE 16
#define NETFLOW1_MAX_FLOWS 24
#define NETFLOW1_FLOW_SIZE 48
#define NETFLOW1_SEQ_OFFSET 0

#define NETFLOW5_VERSION 5
#define NETFLOW5_HEADER_SIZE 24
#define NETFLOW5_MAX_FLOWS 30
#define NETFLOW5_FLOW_SIZE 48
#define NETFLOW5_SEQ_OFFSET 16

#define NETFLOW7_VERSION 7
#define NETFLOW7_HEADER_SIZE 24
#define NETFLOW7_MAX_FLOWS 27
#define NETFLOW7_FLOW_SIZE 52
#define NETFLOW7_SEQ_OFFSET 16

#define NETFLOW8_VERSION 8
#define NETFLOW8_MAX_FLOWS 51

struct NetFlow {
	unsigned Version;
	unsigned HeaderSize;
	unsigned MaxFlows;
	unsigned FlowSize;
	unsigned SeqOffset;
	unsigned HeaderFields;
	uint16_t *HeaderFormat;
	unsigned FlowFields;
	uint16_t *FlowFormat;
};

#endif
